enumerable_methods.rb

```diff
@@ -0,0 +1,147 @@
+#!/usr/bin/ruby
+
+module Enumerable # rubocop:disable Metrics/ModuleLength
+  def my_each
+    return to_enum unless block_given?
+
+    for x in self # rubocop:disable Style/For
+      yield(x)
+    end
+    self
+  end
+
+  def my_each_with_index
+    return to_enum unless block_given?
+
+    i = 0
+    for x in self # rubocop:disable Style/For
+      yield(x, i)
+      i += 1
+    end
+    self
+  end
+
+  def my_select
+    return to_enum unless block_given?
+
+    new_array = []
+    my_each do |x|
+      new_array.push(x) if yield(x)
+    end
+    new_array
+  end
+
+  def my_all?(pat = nil)
+    if block_given?
+      my_each do |x|
+        return false unless yield(x)
+      end
+    else
+      condition = condition_picker(pat)
+      my_each do |x|
+        return false unless condition.call(x)
+      end
+    end
+    true
+  end
+
+  def my_any?(pat = nil)
+    if block_given?
+      my_each do |x|
+        return true if yield(x)
+      end
+    else
+      condition = condition_picker(pat)
+      my_each do |x|
+        return true if condition.call(x)
+      end
+    end
+    false
+  end
+
+  def my_none?(pat = nil)
+    if block_given?
+      my_each do |x|
+        return false if yield(x)
+      end
+    else
+      condition = condition_picker(pat)
+      my_each do |x|
+        return false if condition.call(x)
+      end
+    end
+    true
+  end
+
+  def condition_picker(pat)
+    if pat.is_a?(Class)
+      condition = proc { |x| x.is_a?(pat) }
+    elsif pat.class == Regexp
+      condition = proc { |x| pat === x } # rubocop:disable Style/CaseEquality
+    elsif !pat.nil?
+      condition = proc { |x| x == pat }
+    elsif pat.nil?
+      condition = proc { |x| x }
+    end
+    condition
+  end
+
+  def my_count(item = nil)
+    count = 0
+    if !block_given? || !item.nil?
+      return length unless item
+
+      my_each do |x|
+        count += 1 if x == item
+      end
+      return count
+    end
+    my_each do |x|
+      count += 1 if yield(x)
+    end
+    count
+  end
+
+  def my_map(proc = nil)
+    return to_enum if !block_given? && !proc
+
+    new_array = []
+    my_each do |x|
+      if proc
+        new_array.push proc.call(x)
+      else
+        new_array.push yield(x)
+      end
+    end
+    new_array
+  end
+
+  def my_inject(*args)
+    sym, initial = nil
+    args.each do |x|
+      if x.is_a?(Symbol)
+        sym = x
+        initial ||= 0
+      else
+        initial = x
+      end
+    end
+    initial ||= self[0]
+    if !block_given? && !sym.nil?
+      my_each do |x|
+        initial = x.send(sym, initial)
+      end
+      return initial
+    end
+    my_each do |x|
+      initial = yield(initial, x)
+    end
+    initial
+  end
```

Comment:
- [ ] should be identical and returns the same thing as ruby's [`inject`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-inject). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_inject.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- [ ] my_inject combines all elements of enum by applying a binary operation, specified by a block

Reply:
I believe it's fixed